src/Sets/Zonotope.jl

```diff
@@ -329,30 +329,149 @@ function scale!(α::Real, Z::Zonotope)
 
-- `Z` -- zonotope
-- `r` -- desired order
+    # the first m generators have greatest weight
+    m = floor(Int, n * (r - 1))
 
-### Output
+    # compute interval hull of L
+    Lred = _interval_hull(G, view(indices, (m+1):p))
 
-A new zonotope with fewer generators, if possible.
+    isone(r) && return Zonotope(c, Lred)
 
-### Algorithm
+    Gred = _hcat_KLred(G, view(indices, 1:m), Lred)
+    return Zonotope(c, Gred)
+end
 
-See `overapproximate(Z::Zonotope, ::Type{<:Zonotope}, r::Union{Integer, Rational})`
-for details.
-"""
-function reduce_order(Z::Zonotope, r::Union{Integer, Rational})
-    return overapproximate(Z, Zonotope, r)
+# Implements zonotope order reduction method from [GIR05]
+# We follow the notation from [YS18]
+function _reduce_order(Z::Zonotope, r::Number, ::GIR05)
+    r >= 1 || throw(ArgumentError("the target order should be at least 1, but it is $r"))
+    c = Z.center
+    G = Z.generators
+    n, p = size(G)
+
+    # r is bigger than the order of Z => don't reduce
+    (r * n >= p) && return Z
+
+    # this algorithm sorts generators by ||⋅||₁ - ||⋅||∞ difference
+    indices = Vector{Int}(undef, p)
+    _weighted_gens!(indices, G, _GIR05)
+
+    # the first m generators have greatest weight
+    m = floor(Int, n * (r - 1))
+
+    # compute interval hull of L
+    Lred = _interval_hull(G, view(indices, (m+1):p))
+
+    isone(r) && return Zonotope(c, Lred)
+
+    Gred = _hcat_KLred(G, view(indices, 1:m), Lred)
+    return Zonotope(c, Gred)
+end
+
+# return the indices of the generators in G (= columns) sorted according to the COMB03 method
+# the generator index with highest score goes first
+function _weighted_gens!(indices, G::AbstractMatrix{N}, ::COMB03) where {N}
+    p = size(G, 2)
+    weights = Vector{N}(undef, p)
+    @inbounds for j in 1:p
+        v = view(G, :, j)
+        weights[j] = norm(v, 2)
+    end
+    sortperm!(indices, weights, rev=true, initialized=false)
+    return indices
 end
 
+# return the indices of the generators in G (= columns) sorted according to the GIR05 method
+# the generator index with highest score goes first
+function _weighted_gens!(indices, G::AbstractMatrix{N}, ::GIR05) where {N}
+    n, p = size(G)
+    weights = Vector{N}(undef, p)
+    @inbounds for j in 1:p
+        aux_norm_1 = zero(N)
+        aux_norm_inf = zero(N)
+        for i in 1:n
+            abs_Gij = abs(G[i, j])
+            aux_norm_1 += abs_Gij
+            if aux_norm_inf < abs_Gij
+                aux_norm_inf = abs_Gij
+            end
+        end
+        weights[j] = aux_norm_1 - aux_norm_inf
+    end
+    sortperm!(indices, weights, rev=true, initialized=false)
+    return indices
+end
+
+# compute interval hull of the generators of G (= columns) corresponding to `indices`
+function _interval_hull(G::AbstractMatrix{N}, indices) where {N}
+    n, p = size(G)
+    Lred = zeros(N, n, n)
+    @inbounds for i in 1:n
+        for j in indices
+            Lred[i, i] += abs(G[i, j])
+        end
```

Comment:
```suggestion
    Lred = zeros(N, n, n)
    @inbounds for i in 1:n
        for j in indices
            Lred[i, i] = abs(G[i, j])
        end
```
This is more efficient.